tool/lib/license_utils.dart

```diff
@@ -0,0 +1,255 @@
+          '<$storedName>',
+          defaultStoredValue ?? DateTime.now().year.toString(),
+        );
+        return {
+          existingHeaderKey: existingLicenseText,
+          replacementHeaderKey: replacementLicenseText,
+        };
+      }
+      // Return a non-empty map for the case where there is a stored value
+      // requested (i.e. when there is a '<value>' defined in the license text)
+      final String storedName = _parseStoredName(existingLicenseText);
+      if (storedName.isNotEmpty) {
+        return _processHeaders(
+          storedName,
+          existingLicenseText,
+          replacementLicenseText,
+          content,
+        );
+      }
+    }
+    throw Exception('License header expected in ${file.path}, but not found!');
+  }
+
+  /// Returns a copy of the given [file] with the [existingHeader] replaced by
+  /// the [replacementHeader]. Reads and writes the entire file contents all
+  /// at once, so performance may degrade for large files.
+  Future<File> rewriteLicenseHeader(
+    File file,
+    String existingHeader,
+    String replacementHeader,
+  ) async {
+    final File rewrittenFile = File('${file.path}.tmp');
+    // TODO: [mossmana] Convert to using a stream
+    await file.readAsString().then((String contents) async {
+      final String replacementContents = contents.replaceFirst(
+        existingHeader,
+        replacementHeader,
+      );
+      await rewrittenFile.writeAsString(replacementContents);
+    });
+    return rewrittenFile;
+  }
+
+  /// Bulk update license headers for files in the [directory] as configured
+  /// in the [config]. Returns a list of file paths that were updated.
+  Future<Map<String, List<String>>> bulkUpdate(
+      Directory directory, LicenseConfig config,) async {
+    final List<String> includedPaths = [];
+    final List<String> updatedPaths = [];
+    final List<File> files =
+        directory.listSync(recursive: true).whereType<File>().toList();
+    for (final file in files) {
+      if (!config.shouldExclude(file)) {
+        includedPaths.add(file.path);
+        final String extension = p.extension(file.path).replaceAll('.', '');
+        final YamlList removeIndices =
+            config.getRemoveIndicesForExtension(extension);
+        for (final removeIndex in removeIndices) {
+          final String existingLicenseText = config.removeLicenses[removeIndex];
+          final int addIndex = config.getAddIndexForExtension(extension);
+          final String replacementLicenseText = config.addLicenses[addIndex];
+          final int fileLength = await file.length();
+          final int existingLicenseTextLength = existingLicenseText.length;
+          const int buffer = 20;
+          // Assume that the license text will be near the start of the file,
+          // but add in some buffer.
+          int byteCount = existingLicenseTextLength + buffer;
+          byteCount = fileLength > byteCount ? byteCount : fileLength;
```

Comment:
using the dart:math library, this can be:
`final byteCount = min(existingLicenseTextLength + buffer, fileLength);`